### PR TITLE
Networking v2

### PR DESCRIPTION
**Pull request by magicrub**
This replaces PR https://github.com/ArduPilot/ardupilot/pull/23886 in a much simpler lightweight form. It gives us the ability to ping it and that's about it.

### CODE REVIEW

**Comment by magicrub** [comment]
@tridge and @bugobliterator will be working on changing the memory allocation from being statically assigned to all of SRAM3 (32k) to dynamically allocated when Networking's NET_ENABLED param is enabled. As-is, this PR will cause all H743 and H757 boards to dedicate 32k of RAM to Ethernet, even if Networking is disabled and/or not supported by the hardware.

File: libraries/AP_Networking/AP_Networking.cpp [hunk 1]
```diff
@@ -0,0 +1,393 @@
...
+    // @User: Advanced
+    AP_GROUPINFO("IPADDR0", 1,  AP_Networking,    _param.ipaddr[0],   AP_NETWORKING_DEFAULT_STATIC_IP_ADDR0),
+
+    // @Param: IPADDR1
+    // @DisplayName: IPv4 Address 2nd byte
+    // @Description: Allows setting static IP address. Example: xxx.168.xxx.xxx
+    // @Range: 0 255
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO("IPADDR1", 2,  AP_Networking,    _param.ipaddr[1],   AP_NETWORKING_DEFAULT_STATIC_IP_ADDR1),
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.cpp:44]: What happened to the `4` and `6` for IPV4 vs IPV6 distinction?

File: libraries/AP_Networking/AP_Networking.cpp [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+        AP_HAL::panic("AP_Networking must be singleton");
+    }
+#endif
+
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+
+void AP_Networking::init()
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.cpp:174]: Any reason you made this void return and swallow the error instead of  return a `bool` for success/fail  to propogate the error to the caller?

File: libraries/AP_Networking/AP_Networking.cpp [hunk 3]
```diff
@@ -0,0 +1,393 @@
...
+    hal.gpio->write(HAL_GPIO_ETH_ENABLE, _param.enabled ? 1 : 0);
+#endif
+
+    if (!_param.enabled || _init.done) {
+        return;
+    }
+
+    // set default MAC Address lower 3 bytes to UUID if possible
+    uint8_t uuid[12];
+    uint8_t uuid_len = sizeof(uuid);
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.cpp:187]: Add const?

File: libraries/AP_Networking/AP_Networking.cpp [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+    static char _str_buffer[20];
+    if (hal.util->snprintf(_str_buffer, sizeof(_str_buffer), "%u.%u.%u.%u", (unsigned)ip[0], (unsigned)ip[1], (unsigned)ip[2], (unsigned)ip[3]) == 0) {
+        _str_buffer[0] = '\0';
+    }
+    return _str_buffer;
+}
+char* AP_Networking::convert_ip_to_str(const uint32_t ip)
+{
+    uint8_t ip_array[4];
+        ip_array[3] = ((ip >> 24) & 0xff);
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.cpp:346]: Have you not run the code formatter on this new code? The indent looks off here. You can even enforce it in CI.

File: libraries/AP_Networking/AP_Networking_Config.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+// Below are default params
+// ---------------------------
+
+#ifndef AP_NETWORKING_DEFAULT_OPTIONS
+    #define AP_NETWORKING_DEFAULT_OPTIONS 0
+#endif
+
+// Default DHCP
+#ifndef AP_NETWORKING_DEFAULT_DHCP_ENABLE
+#define AP_NETWORKING_DEFAULT_DHCP_ENABLE 0
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking_Config.h:24]: Have you considered turning DHCP on by default to match behavior of most OS's? An average user might struggle more with their OS-specific network changes than having a common way of adjusting params in ArduPilot that we can document.

File: libraries/AP_Networking/AP_Networking_Config.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+#ifndef AP_NETWORKING_DEFAULT_STATIC_IP_ADDR3
+    #define AP_NETWORKING_DEFAULT_STATIC_IP_ADDR3   14
+#endif
+
+// Default Netmask: 24
+// Note, the netmask is the number of consecutive bits
+#ifndef AP_NETWORKING_DEFAULT_NETMASK
+    #define AP_NETWORKING_DEFAULT_NETMASK       24 // 255.255.255.0
+    // #define AP_NETWORKING_DEFAULT_NETMASK    16 // 255.255.0.0
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking_Config.h:45]: Remove? Netmask 24 LGTM.

File: libraries/AP_Networking/AP_Networking.cpp (hunk 1 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:44]: We're sticking with 4 only for now. Plus, @tridge has a new Param object he wants to implement for quad unit8 for IPs so this stuff will be changing in the future anyway. Also, it's common to have IP meaning v4 and ipv6 for 6 so I think this is fine for now anyway.

File: libraries/AP_Networking/AP_Networking.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:174]: All the inits are like this. The caller doesn't have any error handling. The module itself must handle then

File: libraries/AP_Networking/AP_Networking.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:187]: Sure thing

File: libraries/AP_Networking/AP_Networking.cpp (hunk 4 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:346]: No. I better do that

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 5 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_Config.h:24]: Thought about it.. but chances are pretty low that a DHCP server will be running on the same network as this. As a default it wouldn't hurt though

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 6 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_Config.h:45]: I'd like to keep this in as an example. Maybe I just need to add a comment

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 5 above)
Comment (timtuxworth) [reply libraries/AP_Networking/AP_Networking_Config.h:24]: Want to make a bet?

File: libraries/AP_Networking/AP_Networking.cpp (hunk 3 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:187]: ah! It cannot be const because get_system_id_unformatted() populates the len

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 5 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_Config.h]: Default is now DHCP Enabled

File: libraries/AP_Networking/AP_Networking_Config.h (hunk 6 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking_Config.h]: added more comments

File: libraries/AP_Networking/AP_Networking.cpp (hunk 2 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:174]: For example, here's where it and a bunch of other libraries are calling init at boot
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_Vehicle/AP_Vehicle.cpp#L271

Keep in mind, errors are not treated the same in the embedded world vs desktop OS world

**Comment by bugobliterator** [comment]
@magicrub I have added commits to do dynamic allocation of Ethernet descriptors and buffers. I haven't tested this at all. Feel free to try them. Very likely this will immediately crash, let me know how it goes, I will try it next week as well.

**Comment by bugobliterator** [comment]
Dynamic buffer allocation is tested working now.

**Comment by rmackay9** [comment]
@magicrub,

nicely done!  By the way, I can see some clean up could be done on the commits.  For example there are multiple commits to update the ChibiOS submodule, there's also a revert commit in there.  It should probably also be rebased on master.

Anyway, that's little stuff, really great to see this!

File: libraries/AP_Networking/AP_Networking.cpp [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+    // @Description: Allows setting static IP address. Example: xxx.xxx.xxx.14
+    // @Range: 0 255
+    // @RebootRequired: True
+    // @User: Advanced
+    AP_GROUPINFO("IPADDR3", 4,  AP_Networking,    _param.ipaddr[3],   AP_NETWORKING_DEFAULT_STATIC_IP_ADDR3),
+
+    // @Param: NETMASK
+    // @DisplayName: IP Subnet mask
+    // @Description: Allows setting static subnet mask. The value is a count of consecutive bits. Examples: 24 = 255.255.255.0, 16 = 255.255.0.0
+    // @Range: 0 32
```
Comment (rmackay9) [inline libraries/AP_Networking/AP_Networking.cpp:65]: Can we make this a bitmask or a list of values?

File: Tools/AP_Periph/AP_Periph.cpp [hunk 8]
```diff
@@ -100,6 +100,10 @@ void AP_Periph_FW::init()
 
     can_start();
 
+#if AP_NETWORKING_ENABLED
```
Comment (peterbarker) [inline Tools/AP_Periph/AP_Periph.cpp:103]: Sadly, I believe we're requiring `#ifdef HAL_PERIPH_ENABLE_NETWORKING`.  You will need to turn `AP_NETWORKING_ENABLED` on in `chibios_hwdef.py` if that thing is defined.  Don't blame me, I'm not keen on the rule either.

File: libraries/AP_Networking/AP_Networking.cpp [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+#include <sys/socket.h>
+
+#endif
+
+extern const AP_HAL::HAL& hal;
+
+const AP_Param::GroupInfo AP_Networking::var_info[] = {
+
+    // @Param: ENABLED
+    // @DisplayName: Enable Networking
```
Comment (rmackay9) [inline libraries/AP_Networking/AP_Networking.cpp:23]: In general I think we should put "Networking" at the front of the Display name.  This will make the subsystem clearer for users.

File: libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef-bl.dat [hunk 10]
```diff
@@ -20,6 +20,8 @@ FLASH_RESERVE_START_KB 0
 # the H757 has 128k sectors, assign 2 sectors for BL
 FLASH_BOOTLOADER_LOAD_KB 256
 
+define WATCHDOG_DISABLED
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef-bl.dat:23]: why disable watchdog?

File: libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef.dat [hunk 11]
```diff
@@ -4,9 +4,13 @@
 # MCU class and specific type
 MCU STM32H7xx STM32H757xx
 
+define WATCHDOG_DISABLED
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef.dat:7]: why disable watchdog?

Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef.dat:7]: Seems strange.  Likewise in the bootloader hwdef.

File: libraries/AP_Networking/AP_Networking.cpp [hunk 12]
```diff
@@ -0,0 +1,458 @@
...
+    if (mem == nullptr) {
+        return false;
+    }
+
+    // for total_size == 9240, size should be 16384 and (rasr-1) should be 13 (MPU_RASR_SIZE_16K)
+    const uint32_t rasr_size = MPU_RASR_SIZE(rasr-1);
+
+    // set up MPU region for buffers
+    mpuConfigureRegion(STM32_NOCACHE_MPU_REGION_ETH,
+                       (uint32_t)mem,
```
Comment (tridge) [inline libraries/AP_Networking/AP_Networking.cpp:206]: does this work with the alignment of mem from malloc?

File: libraries/AP_Networking/AP_Networking.h [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+
+class AP_Networking
+{
+public:
+    AP_Networking();
+
+    /* Do not allow copies */
+    CLASS_NO_COPY(AP_Networking);
+
+    void init();
```
Comment (rmackay9) [inline libraries/AP_Networking/AP_Networking.h:32]: I think we should add some more comments above these methods.  Many of the names make it obvious but still, I think we've got many less comments than most libraries.

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py [hunk 14]
```diff
@@ -54,10 +54,10 @@
     # we can't use DTCM first for main firmware as some builds overflow the first segment
     'RAM_MAP_BOOTLOADER' : [
         (0x20000000, 128, 2), # DTCM, tightly coupled, no DMA, fast
-        (0x30000000, 256, 0), # SRAM1, SRAM2
+        (0x30000000, 256, 8), # SRAM1, SRAM2
         (0x24000000, 512, 4), # AXI SRAM. Use this for SDMMC IDMA ops
         (0x00000400,  63, 2), # ITCM (first 1k removed, to keep address 0 unused)
-        (0x30040000,  32, 0), # SRAM3.
+		(0x30040000,  32, 8), # SRAM3.
```
Comment (rmackay9) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py:60]: nitpick: little whitespace issue.  looks like tabs have been added instead of spaces.

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_eth.ld [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+/*
+    this file is included by the board specific ldscript.ld which is
+    generated from hwdef.dat
+*/
+
+SECTIONS
+{
+    /* Special section for Ethernet DMA non cache-able areas.*/
+    .eth (NOLOAD) : ALIGN(4)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/common_eth.ld:25]: why is this needed with the new malloc approach?

File: libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c [hunk 16]
```diff
@@ -371,6 +376,19 @@ void *malloc_axi_sram(size_t size)
...
+/*
+  allocate memory for ethernet DMA
+*/
+void *malloc_eth_safe(size_t size)
+{
+#if defined(STM32H7)
+    return malloc_flags(size, MEM_REGION_FLAG_ETH_SAFE);
+#else
+    (void)size;
+    return NULL;
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c:388]: This stuff testable in SITL?

Do we have any autotests coming for this?

**Review by tridge — CHANGES_REQUESTED** [review]
@bugobliterator can we chat about the memory approach?

File: libraries/AP_HAL_ChibiOS/hwdef/kha_eth/hwdef.dat [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+
+define PERIPH_FW TRUE
+#define NO_DATAFLASH TRUE
+define HAL_LOGGING_ENABLED TRUE
+
+#################################
+# AP_Periph - configurations specific to this App
+#################################
+
+env DISABLE_SCRIPTING 1
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/kha_eth/hwdef.dat:151]: This doesn't work - I think you want `AP_SCRIPTING_ENABLED` ATM.

File: libraries/AP_HAL_ChibiOS/hwdef/kha_eth/hwdef.dat [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+
+# listen for reboot command from uploader.py script
+# undefine to disable. Use -1 to allow on all ports, otherwise serial number index defined in SERIAL_ORDER starting at 0
+define HAL_PERIPH_LISTEN_FOR_SERIAL_UART_REBOOT_CMD_PORT 0
+#################################
+
+
+DMA_NOSHARE USART1* USART2* CAN1*
+
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/kha_eth/hwdef.dat:160]: Did you want this built on the autotest server?

Set AUTOBUILD_TARGETS to None if you don't want this built on autotest.

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py [hunk 19]
```diff
@@ -29,11 +29,11 @@
     # flags of 2 means faster memory for CPU intensive work
     # flags of 4 means memory can be used for SDMMC DMA
     'RAM_MAP' : [
-        (0x30000000, 256, 0), # SRAM1, SRAM2
+        (0x30000000, 256, 8), # SRAM1, SRAM2
         (0x20000000, 128, 2), # DTCM, tightly coupled, no DMA, fast
         (0x24000000, 512, 4), # AXI SRAM. Use this for SDMMC IDMA ops
         (0x00000400,  63, 2), # ITCM (first 1k removed, to keep address 0 unused)
-        (0x30040000,  32, 0), # SRAM3.
+		(0x30040000,  32, 8), # SRAM3.
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py:36]: Indenting weirdness?

Aren't these supposed to be generated?

File: libraries/AP_Networking/AP_Networking.cpp [hunk 20]
```diff
@@ -0,0 +1,458 @@
+
+#include "AP_Networking.h"
+
+#if AP_NETWORKING_ENABLED
```
Comment (peterbarker) [inline libraries/AP_Networking/AP_Networking.cpp:4]:
```suggestion
#include "AP_Networking_config.h"

#if AP_NETWORKING_ENABLED

#include "AP_Networking.h"
```

**Comment by rmackay9** [comment]
Is the AP_NETWORKING_ENABLED define added to the custom build server?  I don't immediately see it.

File: libraries/AP_Networking/AP_Networking.cpp [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+uint32_t AP_Networking::convert_str_to_ip(char* ip_str)
+{
+    uint32_t ip = 0;
+    inet_pton(AF_INET, ip_str, &ip);
+    return ip;
+}
+
+char* AP_Networking::convert_ip_to_str(const uint8_t ip[4])
+{
+    static char _str_buffer[20];
```
Comment (peterbarker) [inline libraries/AP_Networking/AP_Networking.cpp:400]: This should really be passed in.

File: libraries/AP_Networking/AP_Networking.h [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
+    static int32_t send_udp(struct udp_pcb *pcb, const ip4_addr_t &ip4_addr, const uint16_t port, const uint8_t* data, uint16_t data_len);
+#endif
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    static int tftp_fd;
```
Comment (peterbarker) [inline libraries/AP_Networking/AP_Networking.h:153]: This is strange!  Used?

**Review by peterbarker — DISMISSED** [review]

File: libraries/AP_HAL_ChibiOS/system.cpp [hunk 23]
```diff
@@ -304,6 +304,13 @@ void __entry_hook()
 }
 #endif
 
+uint32_t chibios_rand_generate()
+{
+    uint32_t val;
+    hal.util->get_random_vals((uint8_t*)&val, sizeof(val));
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/system.cpp:310]: that function can fail

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py (hunk 14 above)
Comment (Ryanf55) [reply libraries/AP_HAL_ChibiOS/hwdef/scripts/STM32H757xx.py:60]: Feel free to use the `black` formatter which is used in pre-commit to automatically handle these changes.

File: libraries/AP_HAL_ChibiOS/system.cpp (hunk 23 above)
Comment (Ryanf55) [reply libraries/AP_HAL_ChibiOS/system.cpp:310]: The `get_random_vals` has signature:     `virtual bool get_random_vals(uint8_t* data, size_t size) { return false; }`. This is a great example where `WARN_IF_UNUSED` would be helpful at compile time to warn to prevent misuse.

File: libraries/AP_HAL_ChibiOS/hwdef/common/common_eth.ld (hunk 15 above)
Comment (magicrub) [reply libraries/AP_HAL_ChibiOS/hwdef/common/common_eth.ld:25]: Right now the dynamic malloc is a build option depending on STM32_ETH_BUFFERS_EXTERN. Without it this linker file (and changes to the memory map) are needed. If STM32_ETH_BUFFERS_EXTERN is going to be a requirement for Ardupilot in ChibiOS then the linker file is not needed. I'm in favor of removing it

File: libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c (hunk 16 above)
Comment (magicrub) [reply libraries/AP_HAL_ChibiOS/hwdef/common/malloc.c:388]: SITL support will be in a follow-up PR

File: libraries/AP_Networking/AP_Networking.h (hunk 22 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.h:153]: whoops! That's leftover from one of many other variants of this branch where TFTP is supported. Needs to be removed from this version/PR

File: libraries/AP_Networking/AP_Networking.cpp (hunk 21 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:400]: agreed, I'll change it

File: libraries/AP_HAL_ChibiOS/system.cpp (hunk 23 above)
Comment (magicrub) [reply libraries/AP_HAL_ChibiOS/system.cpp:310]: hack pseudo code that isn't full-proof.. but is there such a full-proof solution:
```
WITH_SEMEPHORE()
uint32_t result;
for (uint8_t i=0; i< 10; i++) {
    if (get_random_hard(random_value)) {
        return random_value
    }
    delay_us(1) // or whatever time the RNG requires for various HW targets
}
get_random_soft(random_value)
return random_value
```

File: libraries/AP_Networking/AP_Networking.cpp [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+    uint8_t rasr = 0;
+    // find size closest to power of 2
+    while (size < total_size) {
+        size = size << 1;
+        rasr++;
+    }
+    void *mem = malloc_eth_safe(size);
+    // ensure our memory is aligned
+    // ref. Cortex-M7 peripherals PM0253, section 4.6.4 MPU region base address register
+    if (((uint32_t)mem) % size) {
```
Comment (magicrub) [inline libraries/AP_Networking/AP_Networking.cpp:199]: This belongs after the null check

File: libraries/AP_Networking/AP_Networking.cpp (hunk 21 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.cpp:405]: bleh.. Now I remember why I did it this way. It gets really messy when I do it the other way. I changed it's usage to const char* which is a tiny bit better. Also, note that these are static functions. I think another reason I did it this way was because Scripting likes it better.

File: libraries/AP_Networking/AP_Networking.h (hunk 22 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.h:153]: removed

File: libraries/AP_Networking/AP_Networking.h (hunk 13 above)
Comment (magicrub) [reply libraries/AP_Networking/AP_Networking.h:32]: I've added a bunch.. We could still use more though. There will be plenty of follow-up PRs expanding on all this stuff. I fact, most of them are not currently used in this PR. They're all helpers for things like scripting.

**Review by magicrub — COMMENTED** [review]
I've applied all the suggestions.

**Comment by magicrub** [comment]
> Is the AP_NETWORKING_ENABLED define added to the custom build server? I don't immediately see it.

Whoops, forgot this one

File: libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef.dat (hunk 11 above)
Comment (magicrub) [reply libraries/AP_HAL_ChibiOS/hwdef/CubePilot-CANMod/hwdef.dat]: Fixed

**Review by tridge — CHANGES_REQUESTED** [review]
i've pushed a large pile of changes, ready for testing
 - added classes for IPv4 and MAC address parameters
 - split off ChibiOS backend 
 - fixed memory handling in lwip init code
 - tidied up the default parameter values

**Comment by magicrub** [comment]
Nice refactor!

**Review by tridge — APPROVED** [review]

File: libraries/AP_Networking/AP_Networking.cpp [hunk 25]
```diff
@@ -390,7 +284,7 @@ uint8_t AP_Networking::convert_netmask_ip_to_bitcount(const uint32_t netmask_ip)
     return netmask_bitcount;
 }
 
-uint32_t AP_Networking::convert_str_to_ip(char* ip_str)
+uint32_t AP_Networking::convert_str_to_ip(const char* ip_str)
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking.cpp:164]: Add WARN_IF_UNUSED?

File: libraries/AP_Networking/AP_Networking_address.h [hunk 26]
```diff
@@ -0,0 +1,36 @@
...
+{
+public:
+    AP_Networking_IPV4(const char *default_addr);
+    AP_Int16 addr[4];
+
+    // return address as a uint32_t
+    uint32_t get_uint32(void) const;
+
+    // set address from a uint32_t
+    void set_uint32(uint32_t addr);
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking_address.h:16]: Add `const` to the `addr` param?

File: libraries/AP_Networking/AP_Networking_address.h [hunk 27]
```diff
@@ -0,0 +1,36 @@
...
+
+    // set default address from a uint32
+    void set_default_uint32(uint32_t addr);
```
Comment (Ryanf55) [inline libraries/AP_Networking/AP_Networking_address.h:19]: Add const to the `addr` param?

**Review by Ryanf55 — APPROVED** [review]
I reviewed your changes Tridge, LGTM. Splitting the addresses+params into its own class is much clearer.
